README.md
net2blend introduction
================
Julian Evans
30/11/21

-   [Initial set up](#initial-set-up)
-   [Basic network export](#basic-network-export)
-   [Animating networks](#animating-networks)
    -   [Simple example](#simple-example)
    -   [Animating networks of changing
        size](#animating-networks-of-changing-size)
-   [Animating curved and dashed edges, importing
    folders.](#animating-curved-and-dashed-edges-importing-folders.)
-   [Using custom vertex objects.](#using-custom-vertex-objects.)
-   [Arrows](#arrows)
-   [2d networks in 3d Space](#d-networks-in-3d-space)

<img src="net2blendR/vignettes/examples.jpg" title="" alt="Examples" width="100%" />

Net2Blend is a way of exporting igraph network objects from R and
importing them into blender, allowing the creation of interesting
visualisations of networks and animations of networks.

<img src="net2blendR/vignettes/examplevid1.gif" title="" alt="Example vid1" width="100%" />

<img src="net2blendR/vignettes/examplevid2.gif" title="" alt="Example vid2" width="100%" />

# Initial set up

### Blender

Download the import\_network.py file. Follow the instructions
[here](https://docs.blender.org/manual/en/latest/editors/preferences/addons.html)
for how to install Blender addons. If it works, the “import network” tab
should appear (see below).

### R

The R package can either be installed using the binaries provided on the
github repository, or directly using

``` r
devtools::install_github("jevansbio/net2blend",subdir="net2blendR")
```

We can then load the library. We’ll also load igraph to build our
networks.

``` r
library(igraph)
library(net2blendR)
```

# Basic network export

### R

Here we’ll walk through the most basic export possible.

For this example, we’ll load an artificial network provided with the
package.

``` r
data(examplenet1)
summary(examplenet1)
#> IGRAPH 9632bf4 UN-- 13 21 -- 
#> + attr: name (v/c)
```

This is a simple randomly generated graph. Currently the only attribute
is a name attribute for each vertex. We can now generate a simple circle
layout for this graph.

``` r
l1=as.data.frame(layout_in_circle(examplenet1))
row.names(l1)=V(examplenet1)$name
head(l1)
#>           V1        V2
#> E  1.0000000 0.0000000
#> L  0.8854560 0.4647232
#> M  0.5680647 0.8229839
#> F  0.1205367 0.9927089
#> X -0.3546049 0.9350162
#> N -0.7485107 0.6631227
```

Note that these coordinates are all between -1 and 1. When imported into
Blender these coordinates will be in the scene units, meters by default.
We’ll multiply by 10 to avoid our network being too squashed.

``` r
l1=l1*10
```

We can then export this network using the graph, layout and the default
settings of net2blend. We set the name of the network to be examplenet.

``` r
net2blend(examplenet1,l1,netname="examplenet")
```

This will generate two files. One contains vertex data
“examplenet\_vdata\_0\_.csv” the other edge data
“examplenet\_edata\_0\_.csv”.

### Blender

Open Blender and create a new general scene. Delete the default cube. If
you have installed the net2blend add-on correctly you should find the
“Import network” tab in the side bar (if this is hidden, click the
circled arrow).

<img src="net2blendR/vignettes/blender1.jpg" title="Empty scene" alt="Empty scene" width="100%" />

We’ll use the “import single network” tool to import the network. Set
the edge and vertex files and click the import network button. The
network will be imported.

<img src="net2blendR/vignettes/blender2.jpg" title="Import network" alt="Import network" width="100%" />

You’ll see that 2 collections have been added to the scene.
“examplenetnodes” (yes I’m sorry, I have been inconsistent in my use of
vertices/nodes) and “examplenetedges”. The vertices and edges are found
within these collections.

<img src="net2blendR/vignettes/blender3.jpg" title="Network objects" alt="Network objects" />

You now have a network as a 3d object in Blender, ready to render!

# Animating networks

We’ll now look at how we can animate these networks in Blender. When you
import a network a [keyframe](https://en.wikipedia.org/wiki/Key_frame)
is automatically added for every vertex and edge. Blender will
interpolate between two keyframes, creating a smooth transition. When we
imported our network in the previous example, the “Frame” argument was
set to 0, meaning keyframes were added at frame 0. If we import another
network with the same edges/vertices

We’ll first add simple animation to the existing network and then look
at how to set up networks to animate edges/vertices appearing and
disappearing.

## Simple example

In this example we’ll add some simple changes to the network in our next
keyframe. We’ll keep the network the same, but change the export
settings.

### R

First we’ll generate a new layout. Once again we multiply everything by
10 to put it on a better scale in blender. Since we are working in 3d
space, lets also move the network up by 10 meters.

``` r
l2=as.data.frame(layout_in_circle(examplenet1))
row.names(l2)=V(examplenet1)$name
l2=l2*10
l2$Z = 10 #adding a Z column
```

we’ll export the network with the new layout and different colour vertex
and edges. Importantly, we keep the same network name “examplenet”. This
means that when we import these files, they will edit the existing
networks, either overwriting existing keyframes (if imported at frame 0)
or adding new keyframes (if added at any other frame). We give it a
different netname2, meaning the filenames will have a different suffix
to distinguish it. The filenames will be “examplenet\_vdata\_2\_.csv”
and “examplenet\_edata\_2\_.csv”.

``` r
net2blend(examplenet1,l2,netname="examplenet",netname2=2,edge.color = "red",vertex.color = "blue")
```

### Blender

In blender, change the edge and vertex data paths to the new files. Set
the frame to 30 and click import.

<img src="net2blendR/vignettes/blender4.jpg" title="New keyframe" alt="New keyframe" width="100%" />

If the network seems to disappear, it is because your current viewpoint
isn’t pulled back enough to see the network’s new Z position. The
current frame on the timeline will be set to 30, where a new keyframe
has been added for every node and edge. You can slide the timeline back
and forth and see the network move and change colour.

<img src="net2blendR/vignettes/blendergif1.gif" title="Animated network" alt="Animated network" width="100%" />

## Animating networks of changing size

Lets consider a more complicated example where the number of edges and
nodes change. This requires us to plan ahead a little when exporting our
keyframes.

### R

In order to animate the networks properly, we need all edges and vertex
from all networks to be **in** every network object, even if they are
not actually **present** in that network. This is because we need to
tell Blender whether or not an edge or a vertex needs to not be present
in a particular timestep.

Start with the two example networks. We will first add attributes to
these networks to indicate that an edge or a vertex is definitely
present.

``` r
data(examplenet1)
data(examplenet2)

#give all these nodes an attribute that they were present
V(examplenet1)$present=T
V(examplenet2)$present=T
#give edges a weight attribute to indicate presence
E(examplenet1)$weight=1
E(examplenet2)$weight=1
```

Examplenet2 is larger than examplenet1, and is missing some of the
vertices from examplenet1. We now need to add all possible edges and
nodes to both of these networks. net2blend has a few convenience
functions for this.

``` r
#find all nodes
allvertex=find_all_vertex(list(examplenet1,examplenet2))

#find all edges
alledges=find_all_edges(list(examplenet1,examplenet2))
```

These functions both take a list of network objects and return all the
vertices and edges. We can then add these to every network. We can add
them with certain attributes, so here we can take advantage of the
“present” vertex attribute we added earlier, setting it to FALSE for all
our new vertices.

``` r
#add nodes to g1 with present attribute set to false
examplenet1=add_missing_vertex(examplenet1,allvertex=allvertex,attrlist=list(present=F))

#If we had nodes disappearing in g2, can run the same line
examplenet2=add_missing_vertex(examplenet2,allvertex=allvertex,attrlist=list(present=F))
```

Similarly, we can add all edges to both networks. Here we set the weight
attribute of all our new edges to zero.

``` r
#add missing edges to g1 and g2 with a weight of 0
examplenet1=add_missing_edges(examplenet1,alledges=alledges,attrlist=list(weight=0))
examplenet2=add_missing_edges(examplenet2,alledges=alledges,attrlist=list(weight=0))
```

Both networks should now have the same number of edges and vertices. We
can also make a few extra changes for the sake of readability.

``` r
#For convenience (not required though) can also match the order of nodes
examplenet1=permute(examplenet1,match(V(examplenet1)$name,sort(V(examplenet1)$name)))
examplenet2=permute(examplenet2,match(V(examplenet2)$name,sort(V(examplenet2)$name)))

#generate edge names for convenience - makes it easier to find the same edges in different graphs
E(examplenet1)$name=get_edge_names(examplenet1)
E(examplenet2)$name=get_edge_names(examplenet2)
```

Finally we need to generate a new layout that fits the full network. For
fun, we can once again add some movement on the Z axis.

``` r
#generate a layout for both networks to use. 
#This will be identical, but we'll add a change in z coordinate in the second frame
l1=as.data.frame(layout_in_circle(examplenet1))
row.names(l1)=V(examplenet1)$name
#add a z coordinate
l1$z=0
#make a copy
l2=l1
#add z components 
l2$z[row.names(l2)%in%V(examplenet1)[present]$name]=-0.5
l2$z[!row.names(l2)%in%V(examplenet1)[present]$name]=0.5

#rescale
l1=l1*10
l2=l2*10
```

We are now ready to export both these networks. We use our “present” and
“weight” attribute to set the size of vertices and edges respectively.

``` r
#export examplenet1
net2blend(examplenet1,layout=l1[match(V(examplenet1)$name,row.names(l1)),],#make sure vertices are matched to their coords
vertex.color="red",
edge.color="black",
edge.size=0.1*E(examplenet1)$weight,#edges of weight 0 will be invisible
vertex.size=0.2*V(examplenet1)$present,#non present nodes will be invisible
netname="example1",
netname2=1)
```

Here you can see I am simply multiplying the desired vertex and edge
size by whether or not the edge or vertex actually exists.

We can also use these attributes to animate other aspects of the network
such as colour.

``` r
#for this example, edges that are present in examplenet1 will become blue in examplenet2 while edges that are disappearing will become red
edgecol=rep("black",length(E(examplenet2)))
edgecol[E(examplenet2)$weight==0]="red"
edgecol[E(examplenet2)$name%in%E(examplenet1)[weight>0]$name&E(examplenet2)$weight>0]="blue"

#similarly new nodes will be green while old nodes will become blue.
nodecol=rep("green",length(V(examplenet2)))
nodecol[V(examplenet2)$name%in%V(examplenet1)[present]$name]="blue"
```

We can now export examplenet2 using these settings

``` r
net2blend(examplenet2,layout=l2[match(V(examplenet2)$name,row.names(l2)),],
vertex.color=nodecol,
edge.color=edgecol,
edge.size=0.1*E(examplenet2)$weight,#edges of weight 0 will be invisible
vertex.size=0.2*V(examplenet2)$present,#non present nodes will be invisible
netname="example1",
netname2=2)#note that netname 2 is different
```

### Blender

Open a new scene and import both networks at different frames:

<img src="net2blendR/vignettes/blendergif2.gif" title="animated network 2" alt="animated network 2" width="100%" />

As you can see, edges and nodes now appear and disappear.

# Animating curved and dashed edges, importing folders.

As with edges and vertices disappearing and appearing, a little bit of
forethought is needed to animate curved and dashed edges. If you want to
have an edge curve or become dashed at any point, you need to consider
this in the initial setup.

### R

In order to animate curves and dashes, we export the initial network so
as to set up edges to be curved and dashed, even if they are currently
not. We’ll also save these networks to a folder called example1 and use
netname2 to set frame number. This next code assumes you ran all the
code above so as to make the networks of equal size and set up colours.

``` r
#export examplenet1
net2blend(examplenet1,layout=l1[match(V(examplenet1)$name,row.names(l1)),],#make sure vertices are matched to their coords
vertex.color="red",
edge.color="black",
edge.size=0.1*E(examplenet1)$weight,#edges of weight 0 will be invisible
vertex.size=0.2*V(examplenet1)$present,#non present nodes will be invisible
edge.dash=2,#dash size
edge.isdashed=F,#but no dashes yet.
edge.curve=0,#not curved
edge.forcecurve=T, #but force the edge to allow for curves.
outputdir="example1",
netname="example1",
netname2=0)
```

Using the edge.forcecurve argument force Blender to set up the edges to
allow for curving. Giving a “edge.dash” of greater than 0 OR setting
edge.isdashed to TRUE forces Blender to set up the edge to allow
dashing.

``` r
net2blend(examplenet2,layout=l2[match(V(examplenet2)$name,row.names(l2)),],
vertex.color=nodecol,
edge.color=edgecol,
edge.size=0.1*E(examplenet2)$weight,#edges of weight 0 will be invisible
vertex.size=0.2*V(examplenet2)$present,#non present nodes will be invisible
edge.curve=0,
edge.forcecurve=T, #but force the edge to allow for curves.
edge.dash=2,#dash size
edge.isdashed=T,
outputdir="example1",
netname="example1",
netname2=30)#note that we are setting frames here 

net2blend(examplenet1,layout=l1[match(V(examplenet1)$name,row.names(l1)),],#make sure vertices are matched to their coords
vertex.color="red",
edge.color="black",
edge.size=0.1*E(examplenet1)$weight,#edges of weight 0 will be invisible
vertex.size=0.2*V(examplenet1)$present,#non present nodes will be invisible
edge.curve=2,#edges will curve
edge.forcecurve=T, 
edge.dash=0,#dash size is 0.
edge.isdashed=T,
outputdir="example1",
netname="example1",
netname2=60)
```

We export two more networks, once again sending them to a particular
folder and using netname2 to set the frame number.

We we animate dashes in two different ways. In network 1, dash size is
set to 5, but “isdashed” is set to False. When we move to network 2,
dash size is still 5 but “isdashed” is set to True. This results in the
dashes fading in. When we move to network 3, “isdashed” is still true,
but dash size is set to 0. This results in the dashes increasing in size
until the line becomes solid.

\#\#\#Blender

This time we’ll use the import folder of network function so we don’t
have to import each network individually. Browse to the folder where the
networks were exported. There are now a few options. We can set networks
to be simply ordered by their suffix using the “Order networks by
suffix” button, at a set interval apart as determined by the Frames
option. However, we can also tell Blender to set the frames based on the
suffix, which we will do now, setting our three networks at 0, 30 and 60
frames respectively.

<img src="net2blendR/vignettes/blender5.jpg" title="Import folder" alt="Import folder" width="100%" />

As you can see, the dashes fade in but then scale out.

<img src="net2blendR/vignettes/blendergif3.gif" title="animated network 3" alt="animated network 3" width="100%" />

# Using custom vertex objects.

While net2blend can create spheres, cubes, circles and squares, we can
also use any Blender object as a vertex shape allowing some interesting
effects as can be seen in some of the examples above. Doing this is
fairly straightforward.

### Blender

Create a new scene. In this example we’ll use two custom node shapes, a
cylinder and a [“Suzanne” the Blender monkey
head](https://www.quora.com/Why-has-there-been-a-monkey-head-in-the-add-mesh-section-in-every-version-Blender).
You can see that their names are ‘Cylinder and ’Suzanne’ respectively in
the hierarchy. These are the names we need to use them as custom vertex
shapes.

<img src="net2blendR/vignettes/blender6.jpg" title="Custom objects" alt="Custom objects" width="100%" />

### R

We can now use Suzanne and the cylinder as our vertex shapes. Lets start

CHANGELOG.md
# 09/12/21
## Tweaked R code to allow networks with no edges to be exported/imported. Altered Blender add-on to be able to animate custom node objects with pre-defined materials.

# 07/12/21
## Fixed a small bug in R package vignette compilation.

# 04/12/21 
## Tested with Blender 3.0, made the R functions an R package, improved documentation and many bugfixes.
- Fixed dashes again, but they also now work in eevee! Previously the code was altering the "transmission roughness" instead of alpha. Oops.
- edge.isdashed controls the transparency of dashes while edge.dash controls the size. This was meant to be true last time, but definitely is now
- Fixed arrowheads
- Fixed custom vertex shapes, which I broke when tweaking how vertices are linked to collections.

# 20/04/21 
- Altered code so that added nodes/edges are no longer linked to the scene master collection as well as the network collection.

# 14/04/21 
## Added the following features:
- When importing by folder, you can now sort networks by their suffix (netname2 in R). By default Blender will sort networks by their date created.
- When importing by folder, you can now import networks at frames based on their suffix (netname2 in R), rather than at set intervals.

# 19/11/20
- Fixed bug that would prevent import. Thanks to andyrevell for spotting this!

#17/07/20
## Added the following features:
- Changed the position of interface in Blender. Import network now has its own tab instead of being hidden in "tools".
- Added the ability to import entire folders of networks at set frame intervals.
- Added ability to animate line dashes. Still only visible in cycles renderer unfortunatley.
- Added ability to animate whether an edge is curved or not
- Completely changed how dashed edges work. Previously the material would completely break when an edge was at a certain angle.
- Exposed arguments for how much to offset 2d edges.
- Quite a few bugfixes.

# 26/06/20 
## Added the following features:

- Networks can now be animated! Simply set the frame option to wherever you want the keyframe to be added. If the objects exist already they will be modified with the new values. Size, position and colour of edges can have keyframes associated with them in blender. This has neccesitated the changing of a few things:
- Size of nodes is now in relation to the initial size of blender primities (2m) rather than absolute units. This is to ensure consistency when modifying a nodes size.
- Unfortunatley the previous code where a material was created per colour has had to be retired. Each object now has its own material so that this material can be animated.
- There is now an additional suffix option in the R script (netname2). Blender will ignore this additional name. This means that R will export files as follows "netname_ename_netname2.csv". This allows a user to export different versions of the same network for animation.
- There is now a "directed" argument due to the new way in which the script will attempt to create consistent edge names. In the future I'll modify this to check the igraph object instead. I have not tested animations thorougly with arrows and directed networks yet.
- Added a few utility functions to aid in the setting up of networks for animations, namely to fill a network with the empty edges that might not exist in a particular version of that network, but still need a keyframe.

# 26/08/19 
## Added the following features:

- Edges can now be dashed lines. Giving the edge.dash argument a value will make that edge dashed. Note that at the moment edges will only be dashed in the Cycles renderer.
- Edge and node collections now named based on the filename for easier organisation.
- Flat square node size multiplied by 2 for better parity with circular nodes.
- 2d edges offset slightly below nodes to avoid clipping.
- 2d edges offset slightly in the order of edge plotting, so that as with igraph the last edges in the graph will be on top.
- Added a maxlength argument - can be used to control the amount of curve which is usually based on the longest edge on the network. This could lead to networks with fewer or more edges having slightly different curves to each other. The findmaxlength function can be used to obtain this value from one particular network/layout combination. This value can then be used to force other networks to use the same cuves.


net2blendR/R/net2blendfuncs.R
#' Export network files
#'
#' \code{net2blend} Exports networks in a format for importing into blender
#'
#' @param net igraph network object
#' @param layout Matrix or dataframe of the same number of rows as vertices
#'     in the networks, where first column is vertex x coordinates
#'     and second column is vertex y coordinates. Z coordinates can also be
#'     provided.
#' @param vertex.color Vertex colour as hex or colour name.
#'     Can be single value or vector the same length as the number of vertices
#'     in the network. of Default is "red"
#' @param edge.color Edge colour as hex or colour name. Can be a single value or
#'     a vector of the same length as the number of edges in the network.
#'     Default is "black"
#' @param vertex.shape Name of vertex shape.  Can be single value or vector of
#'     the same length as the number of vertices in the network.
#'     Can be "sphere","cube","circle", "square" or "none".
#'     Alternative, this can be the name of an object in the blender scene.
#'     This object will be duplicated and rescaled as necessary. If the
#'     object has no materials assigned, the object will also be recoloured.
#'     Default is "sphere".
#' @param vertex.size Size of vertex object. Can be single value or vector of
#'     the same length as the number of vertices in the network. Defaults to 0.2.
#' @param edge.size Size of edge. Can be a single value or a vector the same
#'     length as the number of edges in the network. Default is 0.1
#' @param edge.3d Boolean determining if edges will be 3d in blender.
#'     If FALSE will flatten this edge. Can be a single value or a vector the
#'     same length as the number of edges in the network. Default is TRUE.
#' @param edge.curve Amount to curve edges. Can be a single value or a vector of
#'      the same length as the number of edges in the network.
#'      If a single value >0, the amount each edge curves will be scaled by the
#'      length of the longest edge in the layout OR by \code{maxlength} if provided.
#' @param maxlength If provided, a single value to scale \code{edge.curve} by.
#'      Default is NA, meaning the length of the longest edge of the layout is used.
#' @param edge.forcecurve Boolean determining if edges will be forced to be
#'      created as curves in blender, even if the edge is straight
#'      (\code{edge.curve == 0}). Can be a single value or a vector of
#'      the same length as the number of edges in the network. Defaults to FALSE.
#' @param vertex.intersect Boolean. If TRUE edges will meet in the center of
#'     the vertex. If FALSE edges will be shortened by size of vertex or by
#'     \code{vertex.edgeshorten}. Can be a single value or a vector of
#'      the same length as the number of vertices in the network.
#'      Default is TRUE.
#' @param vertex.edgeshorten If provided, edges will be shortened by this value
#'     instead of the size of the vertex. Can be a single value or a vector of
#'      the same length as the number of edges in the network. Default is 0.
#' @param edge.arrows Boolean. If TRUE, arrowheads are added at the end of
#'     edges. Can be a single value or a vector of the same length as the number
#'     of edges in the network. Default is FALSE.
#' @param edge.arrowsize Thickness of arrowheads. Can be a single value or a vector of
#'      the same length as the number of edges in the network.
#'      If \code{edge.arrows} is TRUE and no value is provided, defaults to
#'       \code{edge.size + 0.05}.
#' @param edge.arrowlength Length of arrowheads. Can be a single value or a vector of
#'      the same length as the number of edges in the network.
#'      If \code{edge.arrows} is TRUE and no value is provided, this defaults
#'      to 0.2.
#' @param edge.dash Size of dashed edges. Can be a single value or a vector of
#'      the same length as the number of edges in the network.
#'      Larger values result in a greater number of short dashes. Defaults to 0.
#' @param edge.isdashed Boolean determining if edges will be forced to be set up
#'     in Blender to allow dashed edges, even if \code{edge.dash == 0}
#' @param zoffset1 Minimum amount to offset 2d edges (\code{edge.3d == FALSE})
#'     below the Z axis by. Default is 0.01.
#' @param zoffset2 Maximum amount to offset 2d edges (\code{edge.3d == FALSE})
#'     below the Z axis by. Default is 0.05.
#' @param outputdir Directory in which to export the networks. Defaults to NA,
#'     the working directory.
#' @param netname Network prefix name. Will be used as the network name in
#'     Blender. Default is "net"
#' @param netname2 Network suffix name. Can be used as frame number by blender.
#'     Default is 0.
#' @return Exports two files, one for edges (edata) and one for vertices (vdata).
#'     Files are named in the format netname_filetype_netname2.csv. See vignette
#'     for full R to Blender workflow and examples.
#'
#' @examples
#' data(examplenet1)
#' l1=layout_nicely(examplenet1)
#' l1=l1*10
#' net2blend(examplenet1,l1,netname="examplenet")
#' @export
net2blend=function(net,layout,vertex.color="red",edge.color="black",
                   vertex.shape="sphere",	vertex.size=0.2,edge.size=0.1,
                   edge.3d=T,edge.curve=0,edge.forcecurve=F,maxlength=NA,
                   vertex.intersect=T,vertex.edgeshorten=0,edge.arrows=F,
                   edge.arrowsize=0, edge.arrowlength=0,edge.dash=0,
                   edge.isdashed=F,zoffset1=0.01,zoffset2=0.05,	outputdir=NA,
                   netname="net",netname2=0){
	directed=igraph::is.directed(net)

	if(is.na(outputdir)){
		outputdir=getwd()
	}
	if(all(edge.arrows)&all(edge.arrowsize==0)){
		edge.arrowsize=edge.size+0.05
	}
	if(all(edge.arrows)&all(edge.arrowlength==0)){
		edge.arrowlength=0.2
	}

	if((!vertex.intersect&vertex.edgeshorten==0)){
		vertex.edgeshorten=vertex.size-0.05
	}

	if(edge.arrows){
	  vertex.edgeshorten=vertex.edgeshorten+edge.arrowlength
	}

	colnames(layout)=c("x","y","z")[1:ncol(layout)]
	layout=as.data.frame(layout)
	if(!"z" %in% names(layout)){
		layout$z=0
	}
	igraph::V(net)$x=layout$x
	igraph::V(net)$y=layout$y
	igraph::V(net)$z=layout$z

	if(any(edge.curve%in%c(T,F))){
		edge.curve2=edge.curve
		edge.curve=rep(0,length(edge.curve))
		edge.curve[edge.curve2]=0.2
	}

	edge.dash2=edge.dash
	if(any(edge.dash>0)|edge.isdashed){
		edata=igraph::as_long_data_frame(net)

		if(length(edge.dash2)<nrow(edata)){
			edge.dash2=rep(edge.dash2,nrow(edata))
		}
		lengths=sqrt((edata$to_x-edata$from_x)^2+(edata$to_y-edata$from_y)^2+(edata$to_z-edata$from_z)^2)
		edge.dash2[edge.dash>0]=edge.dash2[edge.dash>0]*lengths[edge.dash>0]
	}

	if(length(edge.curve)==1){
		#if only one curve value is supplied, scale by edge length
		edata=igraph::as_long_data_frame(net)
		if(nrow(edata>0)){
		  lengths=sqrt((edata$to_x-edata$from_x)^2+(edata$to_y-edata$from_y)^2+(edata$to_z-edata$from_z)^2)
		  if(is.na(maxlength)){
			  maxlength=max(lengths)
		  }
		  edge.curve=edge.curve*(lengths/maxlength)
		}
	}

	igraph::V(net)$colour=vertex.color
	igraph::E(net)$colour=edge.color

	igraph::V(net)$shape=vertex.shape
	igraph::V(net)$size=vertex.size
	igraph::V(net)$size[igraph::V(net)$shape=="square"]=igraph::V(net)$size[igraph::V(net)$shape=="square"]*2
	igraph::V(net)$shorten=vertex.edgeshorten


	igraph::E(net)$size=edge.size
	igraph::E(net)$curve=edge.curve
	igraph::E(net)$forcecurve=edge.forcecurve
	igraph::E(net)$is3d=edge.3d
	igraph::E(net)$arrowsize=edge.arrowsize
	igraph::E(net)$arrowlength=edge.arrowlength
	igraph::E(net)$dash=edge.dash2
	igraph::E(net)$isdashed=edge.isdashed

	#It is important to permute the nodes to the same order, otherwise the from and to of edges can change, leading to your edges flipping
	net=permute(net,match(igraph::V(net)$name,sort(igraph::V(net)$name)))

	edata=igraph::as_long_data_frame(net)
	if(nrow(edata>0)){
  	if(any(!edata$is3d)){
  		highz=max(c(edata$to_z[!edata$is3d],edata$from_z[!edata$is3d]))-zoffset1
  		lowz=max(c(edata$to_z[!edata$is3d],edata$from_z[!edata$is3d]))-zoffset2

  		edata$to_z[!edata$is3d]=seq(lowz,highz,length.out=sum(!edata$is3d))
  		edata$from_z[!edata$is3d]=seq(lowz,highz,length.out=sum(!edata$is3d))
  	}

  	edata=data.frame(edata,t(grDevices::col2rgb(edata$colour)/255))

  	edata$name=sapply(1:nrow(edata),function(x){
  		cnodes=c(edata$from_name[x],edata$to_name[x])
  		if(!directed){
  			cnodes=cnodes[order(cnodes)]
  		}
  		paste(cnodes,collapse="_")
  	})
	}else{
	  emptynames=c(names(edata),"red","green","blue","name")
    edata=data.frame(matrix(nrow=0,ncol=length(emptynames)))
    colnames(edata)=emptynames
  }
	write.csv(edata,file.path(outputdir,paste(netname,"edata",netname2,".csv",sep="_")),row.names=F)

	vdata=igraph::as_data_frame(net,what = 'vertices')
	vdata=data.frame(vdata,t(grDevices::col2rgb(vdata$colour)/255))
	write.csv(vdata,file.path(outputdir,paste(netname,"vdata",netname2,".csv",sep="_")),row.names=F)

}


#' Find longest edge in a layout
#'
#' \code{findmaxlength} Calculate longest edge in a given layout
#'
#' @param net igraph network object
#' @param layout Matrix or dataframe of the same number of rows as vertices
#'     in the networks, where first column is vertex x coordinates
#'     and second column is vertex y coordinates. Z coordinates can also be
#'     provided.
#' @return Returns the length of the longest edge in a given layout. This can
#'     then be used as the \code{maxlength} argument in \code{net2blend} to keep
#'     curved edges consistent over multiple networks.
#'
#' @examples
#' data(examplenet1)
#' l1=layout_nicely(examplenet1)
#' l1=l1*10
#' findmaxlength(examplenet1,l1)
#' @export
findmaxlength=function(net,layout){

	colnames(layout)=c("x","y","z")[1:ncol(layout)]
	layout=as.data.frame(layout)
	if(!"z" %in% names(layout)){
		layout$z=0
	}
	V(net)$x=layout$x
	V(net)$y=layout$y
	V(net)$z=layout$z
	edata=igraph::as_long_data_frame(net)
	lengths=sqrt((edata$to_x-edata$from_x)^2+(edata$to_y-edata$from_y)^2+(edata$to_z-edata$from_z)^2)
	return(max(lengths))
}

#' Find all possible edges in a list of networks
#'
#' \code{find_all_edges} Find all edges that exist in a list of networks.
#'
#' @param allnets list of igraph network objects
#' @return Returns a dataframe of all edges that exist within \code{allnets}.
#'     This can be used as an argument in \code{add_missing_edges} to ensure
#'     that all edges are correctly generated for animation in Blender.
#'
#' @examples
#' data(examplenet1)
#' data(examplenet2)
#' netlist=list(examplenet1,examplnet2)
#' find_all_edges(netlist)
#' @export
find_all_edges=function(allnets){
  directed=any(sapply(allnets,igraph::is.directed))
	alledges1=sapply(1:length(allnets),function(i){

		allnet=allnets[[i]]
		edges=igraph::as_edgelist(allnet)

		return(edges)
	})
	alledges1=do.call(rbind,alledges1)
	if(!directed){
		alledges1=t(sapply(1:nrow(alledges1),function(x){alledges1[x,order(alledges1[x,])]}))
	}
	alledges2=sapply(1:nrow(alledges1),function(x){paste(alledges1[x,],collapse="_")})
	alledges3=alledges1[!duplicated(alledges2),]
	return(alledges3)
}

#' Add missing edges to a network.
#'
#' \code{add_missing_edges} Add edges to a network to ensure consistency.
#'
#' @param net igraph network object
#' @param alledges Dataframe of edges as generated by \code{find_all_edges}
#' @param attrlist list of attributes to be given to the new edges, where the name
#'     of each list item is the name of an edge attribute.
#' @return Returns a network with the edges of \code{alledges} that are not
#'     present in \code{net} added. The new edges will have the attributes
#'     provided by \code{attrlist}.
#'
#' @examples
#' data(examplenet1)
#' data(examplenet2)
#' netlist=list(examplenet1,examplnet2)#list of nets
#' alledges=find_all_edges(netlist)#get all edges
#' allvertices=find_all_vertex(netlist)#get all vertices
#' #add missing vertices and edges to a network
#' examplenet1=add_missing_vertex(examplenet1,allvertices)
#' examplenet1=add_missing_edgeS(examplenet1,alledges)
#' @export
add_missing_edges=function(net,alledges=NULL,attrlist=list()){
  directed=igraph::is.directed(net)
	if(is.null(alledges)){
		alledges=t(combn(1:length(V(net)),2))
		if(directed){
			alledges=rbind(alledges,alledges[,c(2,1)])
		}
	}
	alledges2=sapply(1:nrow(alledges),function(x){paste(alledges[x,],collapse="_")})

	selfloop=F# no function currently
	if(!selfloop){
		samebool=alledges[,1]!=alledges[,2]
		alledges=alledges[samebool,]
		alledges2=alledges2[samebool]
	}

	curredges=igraph::as_edgelist(net)
	curredges2=sapply(1:nrow(curredges),function(x){paste(curredges[x,],collapse="_")})
	curredges3=sapply(1:nrow(curredges),function(x){paste(curredges[x,c(2,1)],collapse="_")})
	if(directed){
		missingedges=alledges[!alledges2%in%curredges2,]
	}else{
		missingedges=alledges[(!alledges2%in%curredges2)&(!alledges2%in%curredges3),]
	}

	net2=igraph::add_edges(net,t(missingedges),attr=attrlist)

	return(net2)
}

#' Find all possible vertices in a list of networks
#'
#' \code{find_all_vertex} Find all edges that exist in a list of networks.
#'
#' @param allnets list of igraph network objects
#' @return Returns a vector of all vertices that exist within \code{allnets}.
#'     This can be used as an argument in \code{add_missing_vertex} to ensure
#'     that all vertices are correctly generated for animation in Blender.
#'
#' @examples
#' data(examplenet1)
#' data(examplenet2)
#' netlist=list(examplenet1,examplnet2)#list of nets
#' allvertices=find_all_vertex(netlist)#get all vertices
#' @export
find_all_vertex=function(allnets){
	unique(unlist(sapply(allnets,function(x){
	  igraph::V(x)$name
	})))
}

#' Add missing vertices to a network.
#'
#' \code{add_missing_vertices} Add edges to a network to ensure consistency.
#'
#' @param net igraph network object
#' @param allvertex Vector of edges as generated by \code{find_all_vertex}
#' @param attrlist list of attributes to be given to the new vertices, where the name
#'     of each list item is the name of an vertex attribute.
#' @return Returns a network with the vertices of \code{allvertex} that are not
#'     present in \code{net} added. The new vertices will have the attributes
#'     provided by \code{attrlist}.
#'
#' @examples
#' data(examplenet1)
#' data(examplenet2)
#' netlist=list(examplenet1,examplnet2)#list of nets
#' alledges=find_all_edges(netlist)#get all edges
#' allvertices=find_all_vertex(netlist)#get all vertices
#' #add missing vertices to examplenet1
#' examplenet1=add_missing_vertex(examplenet1,allvertices)
#' @export
add_missing_vertex=function(net,allvertex,attrlist=list()){
	if(!"name"%in%attrlist){
		attrlist$name=allvertex[!allvertex%in%V(net)$name]
	}

	net2=igraph::add_vertices(net,length(allvertex[!allvertex%in%igraph::V(net)$name]),attr=attrlist)

	return(net2)
}


#' Get edge names
#'
#' \code{get_edge_names} Utility function to assign each edge a name.
#'
#' @param net igraph network object
#' @return Returns a vector of edge names in the format "vertex1_vertex2"
#' @examples
#' data(examplenet1)
#' get_edge_names(examplenet1)
#' @export
get_edge_names=function(net){
  directed=igraph::is.directed(net)
	edges=igraph::as_edgelist(net)

	if(!directed){
		edges=t(sapply(1:nrow(edges),function(x){edges[x,order(edges[x,])]}))
	}
	edges=sapply(1:nrow(edges),function(x){paste(edges[x,],collapse="_")})
	return(edges)
}

